app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt

```diff
@@ -459,22 +468,71 @@ class WebViewActivity : AppCompatActivity(), io.homeassistant.companion.android.
     }
 
     @SuppressLint("InflateParams")
-    override fun authenticationDialog(handler: HttpAuthHandler) {
+    fun authenticationDialog(handler: HttpAuthHandler, host: String, authError: Boolean) {
         val inflater = layoutInflater
         val dialogLayout = inflater.inflate(R.layout.dialog_authentication, null)
         val username = dialogLayout.findViewById<EditText>(R.id.username)
         val password = dialogLayout.findViewById<EditText>(R.id.password)
-
-        AlertDialog.Builder(this)
-            .setTitle(R.string.auth_request)
-            .setView(dialogLayout)
-            .setPositiveButton(android.R.string.ok) { _, _ ->
-                handler.proceed(username.text.toString(), password.text.toString())
-            }
-            .setNeutralButton(android.R.string.cancel) { _, _ ->
-                Toast.makeText(applicationContext, R.string.auth_cancel, Toast.LENGTH_SHORT).show()
+        val remember = dialogLayout.findViewById<CheckBox>(R.id.checkBox)
+        val viewPassword = dialogLayout.findViewById<ImageView>(R.id.viewPassword)
+        var httpAuthList: HashSet<String> = hashSetOf()
+        var autoAuth = false
+
+        viewPassword.setOnClickListener() {
+            if (password.transformationMethod == PasswordTransformationMethod.getInstance()) {
+                password.transformationMethod = HideReturnsTransformationMethod.getInstance()
+                viewPassword.setImageResource(R.drawable.ic_visibility_off)
+                password.setSelection(password.text.length)
+            } else {
+                password.transformationMethod = PasswordTransformationMethod.getInstance()
+                viewPassword.setImageResource(R.drawable.ic_visibility)
+                password.setSelection(password.text.length)
             }
-            .show()
+        }
+
+        if (presenter.getHttpAuthList() != emptySet<String>()) {
+            httpAuthList = (presenter.getHttpAuthList() as HashSet<String>)
+            if (!authError)
+                for (i in 0 until httpAuthList.count()) {
+                    val httpAuth = httpAuthList.elementAt(i).toString().split("|")
+                    if (httpAuth.elementAt(0).toString().contentEquals(host)) {
+                        handler.proceed(httpAuth.elementAt(1).toString(), httpAuth.elementAt(2).toString())
+                        autoAuth = true
+                        firstAuthTime = System.currentTimeMillis()
+                    }
+                }
+        }
+
+        if (!autoAuth || authError) {
+            AlertDialog.Builder(this)
+                .setView(dialogLayout)
+                .setPositiveButton(android.R.string.ok) { _, _ ->
+                    if (username.text.toString() != "" && password.text.toString() != "") {
+                        val httpAuth = host + "|" + username.text.toString() + "|" + password.text.toString()
```

Comment:
This feels like a gigantic hack for persisting this information.  We should consider persisting this in a full DB rather than 
a Shared Preference.

Reply:
Persisting data is now save in SQLite database, and no longer in Shared Preference.